trollflow2/plugins/__init__.py

```diff
@@ -675,3 +675,72 @@ def _get_plugin_conf(product_list, path, defaults):
         conf[key] = get_config_value(product_list, path, key,
                                      default=defaults.get(key))
     return conf
+
+
+def check_valid(job):
+    """Remove products that have too much invalid data.
+
+    Remove any products where the fraction valid_data/expected_valid_data is
+    less than a configured threshold in %.  Expected valid data is calculated
+    by the scene coverage for each resampled scene.  This plugin was designed
+    for use with AVHRR, which may alternate between channels 3A and 3B.
+    Since this is different between resampled scenes, this plugin must be
+    applied after scene resampling.
+
+    This will trigger a calculation for the data to be checked.
+
+    To be configured with the ``rel_valid`` key indicating validity in %.
+    For example:
+
+        product_list:
+          areas:
+            fribbulus_xax:
+              red:
+                min_valid: 40
+
+        workers:
+          - fun: !!python/name:trollflow2.plugins.create_scene
+          - fun: !!python/name:trollflow2.plugins.load_composites
+          - fun: !!python/name:trollflow2.plugins.resample
+          - fun: !!python/name:trollflow2.plugins.check_valid
+          - fun: !!python/name:trollflow2.plugins.save_datasets
+
+    """
+    exp_cov = {}
+    for (area_name, area_props) in job["product_list"]["product_list"]["areas"].items():
+        to_remove = set()
+        for (prod_name, prod_props) in area_props["products"].items():
+            if "min_valid" in prod_props:
+                LOG.debug(f"Checking validity for {area_name:s}/{prod_name:s}")
+                if prod_name not in job["resampled_scenes"][area_name]:
+                    LOG.debug(f"product {prod_name!s} not found, already removed or loading failed?")
+                    continue
+                prod = job["resampled_scenes"][area_name][prod_name]
+                platform_name = prod.attrs["platform_name"]
+                start_time = prod.attrs["start_time"]
+                end_time = prod.attrs["end_time"]
+                sensor = prod.attrs["sensor"]
+                if area_name not in exp_cov:
+                    # get_scene_coverage uses %, convert to fraction
+                    exp_cov[area_name] = get_scene_coverage(
+                        platform_name, start_time, end_time, sensor, area_name)/100
+                valid = job["resampled_scenes"][area_name][prod_name].notnull()
+                exp_valid = exp_cov[area_name]
+                actual_valid = float(valid.sum()/valid.size)
```

Comment:
This line is slow, because it computes the data.  I warn for this in the documentation.  However, it appears the computation is not persistent: `save_datasets()` is as slow as before, and overall processing is therefore much slower than before.  How can I make sure dask remembers this computation?

Reply:
No longer relevant; everything is persisted at once in a dedicated function.